src/client.js

```diff
@@ -0,0 +1,351 @@
+ * objects in the same process at the same time, while one is being "retired" but
+ * is lingering around to send ack's for any in-flight message handlers.
+ *
+ * The instance's `amqp` property is the amqp connection object.  In the event of any
+ * issues with the connection, call the instance's `failed` method.  This will initiate
+ * a retirement of the connection and creation of a new connection.
+ *
+ * The instance will emit a `connected` event when it connects to the pulse server.
+ * This event occurs before the connection is provided to a user, so it is only
+ * of interest to the Client class.
+ *
+ * This instance will emit a `retiring` event just before it is retired.  Users
+ * should cancel consuming from any channels, as a new connection will soon
+ * begin consuming.  Errors from such cancellations should be logged and
+ * ignored.  This connection will remain open for 30 seconds to allow any
+ * in-flight message processing to complete.
+ *
+ * The instance will emit `finished` when the connection is finally closed.
+ *
+ * A connection's state can be one of
+ *
+ *  - waiting -- waiting for a call to connect() (for minReconnectionInterval)
+ *  - connecting -- waiting for a connection to complete
+ *  - connected -- connection is up and running
+ *  - retiring -- in the process of retiring
+ *  - finished -- no longer connected
+ *
+ *  Note that an instance that fails to connect will skip from `connecting` to
+ *  `retiring`.
+ *
+ */
+class Connection extends events.EventEmitter {
+  constructor(client, id) {
+    super();
+
+    this.client = client;
+    this.id = id;
+    this.amqp = null;
+
+    this.debug = debug(`taskcluster-lib-pulse:connection:${client.id}.${id}`);
+
+    this.debug('waiting');
+    this.state = 'waiting';
+  }
+
+  connect() {
+    if (this.state !== 'waiting') {
+      return;
+    }
+
+    this.debug('connecting');
+    this.state = 'connecting';
+
+    amqplib.connect(this.client.connectionString, {
+      heartbeat: 120,
+      noDelay: true,
+      timeout: 30 * 1000,
+    }).then(
+      amqp => {
+        if (this.state !== 'connecting') {
+          // we may have been retired already, in which case we do not need this
+          // connection
+          amqp.close();
+          return;
+        }
+        this.amqp = amqp;
+
+        amqp.on('error', err => {
+          if (this.state === 'connected') {
+            this.debug(`error from aqplib connection: ${err}`);
+            this.failed();
+          }
+        });
+
+        amqp.on('close', err => {
+          if (this.state === 'connected') {
+            this.debug('connection closed unexpectedly');
+            this.failed();
+          }
+        });
+
+        this.debug('connected');
+        this.state = 'connected';
+        this.emit('connected');
+      },
+      err => {
+        this.debug(`Error while connecting: ${err}`);
+        this.failed();
+      });
+  }
+
+  failed() {
```

Comment:
I think this is wrong. connection should only be considered dead if it throws `on('error', ..)`

Reply:
when amqp throws error, failed() is called. Do u want something like connection.emit('error',failed)

Reply:
Recycling a connection shouldn't be a big deal.  The idea here is to exercise an abundance of caution, rather than relying only on the `error` event at the `connection` level.  Whether to consider any error from a channel as fatal to the connection is debatable.

Reply:
i think  if its only a channel error it may not be fatal to the connection, suppose `source`(exchange) is invalid for `channel.bindQueue` then  if that's only a channel error then that channel should be closed but not entire connection and recycled.

Reply:
Does that produce a channel error?  We need to understand better when the channel emits an `error` event, and when it throws an error from a method.  Does it always do both?

Failing to handle error events results in termination of the process, so we need to handle them *somehow*.  Whether that's ignoring them (because the caller of some method got an error) or recycling the connection, or trying to re-create the channel (but carefully to not create two channels at the same time), I'm not sure.

Perhaps some experimentation with the API is in order:
 * Call a method with the wrong arguments
 * Call a method on a channel attached to a closed connection (for example, after killing the rabbitmq process)

Reply:
I ran a cloudamqp server and observed that

- an error in channel closes that channel but other channels in the connection are not harmed. That is to say no connection close or error event occurs for that.

Further documentation says, well should have been seen that part before

Connection errors :
- a protocol transgression the server detected (likely a bug in this library)
- a server error
- a network error
- the server thinks the client is dead due to a missed heartbeat

Channel errors:
- an operation failed due to a failed precondition (usually something named in an argument not existing)
- a human closed the channel with an admin tool
Further a channel will not emit 'error' if its connection closes with an error.

Logically, if the connection close event occurs, first channels are closed followed by the connection...